# 19.01.11_Call_by_object_reference

> Is python call by value or call by reference?

## Call_by_value

> 어떤 경우 파이썬은 call by value 처럼 보인다.

1. 코드 예시

   ```python
   def test(a):
       a += 1
   
   a = 0
   test(a)
   print(a)
   ```

2. 실행 결과

   ```python
   0
   ```

## Call_by_reference

> 하지만 어떤 경우에서 python은 call by reference 처럼 동작한다.

1. 코드 예시

   ```python
   def test(a):
       a.append("bear")
   
   b = ["teddy"]
   test(b)
   b
   ```

2. 실행 결과

   ```python
   ['teddy', 'bear']
   ```

## Call_by_object_reference

1. 코드 예시[^출처1]

   ```python
   >>> global_immutable = 10
   >>> def test1(arg_immutable):
   ...     print(id(global_immutable), id(arg_immutable))
   ...     print(global_immutable is arg_immutable)
   ...     arg_immutable += 1000
   ...     print(id(global_immutable), id(arg_immutable))
   ...     print(global_immutable is arg_immutable)
   ...
   >>> test1(global_immutable)
   (140275802476416, 140275802476416)
   True
   (140275802476416, 140275802485448)
   False
   
   
   >>> global_mutable = [1, 2, 3, 4]
   >>> def test2(arg_mutable):
   ...     print(id(global_mutable), id(arg_mutable))
   ...     print(global_mutable is arg_mutable)
   ...     arg_mutable += [5, 6, 7]
   ...     print(id(global_mutable), id(arg_mutable))
   ...     print(global_mutable is arg_mutable)
   ...
   >>> test2(global_mutable)
   (4531210648, 4531210648)
   True
   (4531210648, 4531210648)
   True
   ```

   python에는 immutable한 변수형이 있고, mutable한 변수형이 있다. 변수는 그들의 reference만 가지고 있다.  immutable한 변수형은 내용의 변경이 불가능하다. 따라서 새 값을 만들고 reference가 다른 값으로 변경된다.

2. ### 결론

   - python은 **call by reference**로 reference를 넘겨줍니다. 하지만 이때 넘겨주는 것은 변수의 reference가 아니라 변수가 담고 있는 **자료의 reference**다

   - ##### 따라서 자료가 mutable 하다면 자료를 변경해도 reference가 보존 되므로 call by reference처럼 보일 것이고 immutable 하다면 call by value가 될 것이다.



----------------------------------------------------

[^출처1]:https://item4.github.io/2015-07-18/Some-Ambiguousness-in-Python-Tutorial-Call-by-What/